scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -475,35 +490,42 @@ object Execution {
 
   private[scalding] final case class OnComplete[T](prev: Execution[T], fn: Try[T] => Unit) extends Execution[T] {
     protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      Trampoline.call(prev.runStats(conf, mode, cache)).map { res =>
+      Trampoline.call(prev.runStats(conf, mode, cache)).map { case (fut, cancelHandler) =>
         cache.getOrElseInsert(conf, this, {
           /**
            * The result we give is only completed AFTER fn is run
            * so callers can wait on the result of this OnComplete
            */
           val finished = Promise[(T, Map[Long, ExecutionCounters])]()
-          res.onComplete { tryT =>
+          fut.onComplete { tryT =>
             try {
               fn(tryT.map(_._1))
             } finally {
               // Do our best to signal when we are done
               finished.complete(tryT)
             }
           }
-          finished.future
+          (finished.future, cancelHandler)
         })
       }
   }
 
   private[scalding] final case class RecoverWith[T](prev: Execution[T], fn: PartialFunction[Throwable, Execution[T]]) extends Execution[T] {
```

Comment:
I haven't look in detail at this, but I think we don't want to recover cancellations. We may need a special exception and pattern match and make sure we never pass that to fn.

Reply:
👍 added this logic and added some tests